Added "function" capabilities

Eduardo-T opened the pull request:
This fork adds the new function capabilities. Check it out.

AESIRZZW commented:
@TheoKanning Help review this change when you have time, thanks in advance.

AndrewCPU commented on README.md:
```diff
@@ -1,12 +1,79 @@
+
+# 🚀 Publishing to Local Maven Repository (⚠ Unofficial Version with Function Support)
```
This should be removed before merge

AndrewCPU commented on api/build.gradle:
```diff
@@ -2,7 +2,7 @@ apply plugin: 'java-library'
 
 dependencies {
-    implementation 'com.fasterxml.jackson.core:jackson-annotations:2.9.0'
+    api 'com.fasterxml.jackson.core:jackson-annotations:2.9.0'
```
Why does this have to change?

Eduardo-T replied:
It is also being used by the 'service' module.

AndrewCPU commented on applyLocalPublish.gradle:
```diff
@@ -0,0 +1,6 @@
+// applyLocalPublish.gradle
+allprojects {
```
Is this needed?

AndrewCPU commented on localPublish.gradle:
```diff
@@ -0,0 +1,16 @@
+apply plugin: 'maven-publish'
+
```
Is this needed?

AndrewCPU commented on service/build.gradle:
```diff
@@ -5,7 +5,8 @@ dependencies {
     implementation 'com.squareup.retrofit2:adapter-rxjava2:2.9.0'
-    implementation 'com.squareup.retrofit2:converter-jackson:2.9.0'
+    implementation 'com.squareup.retrofit2:converter-jackson:2.7.2'
+    implementation 'com.kjetland:mbknor-jackson-jsonschema_2.13:1.0.39'
```
?

AndrewCPU replied:
This build of com.kjetland:mbknor-jackson-jsonschema_2.13:1.0.39 is very vulnerable and sus?
Why would you downgrade the jackson converter to 2.7.2?
https://mvnrepository.com/artifact/com.kjetland/mbknor-jackson-jsonschema_2.13/1.0.39

AndrewCPU commented on publishToLocal.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+./gradlew -I applyLocalPublish.gradle publishToMavenLocal
```
?

AndrewCPU commented on publishToLocal.bat:
```diff
@@ -0,0 +1,2 @@
+@echo off
```
?

AndrewCPU reviewed (changes requested):
Not a full review, but the unnecessary changes stemming from the local maven repo tutorial should be removed..

Eduardo-T commented:
Thanks @AndrewCPU,

The dependency downgrade was a mistake, I fixed it. And mbknor-jackson-jsonschema_2.13:1.0.39 was changed to mbknor-jackson-jsonschema_2.12:1.0.39 fixing its vulnerabilities. It is needed for the API to work ([as seen here](https://platform.openai.com/docs/api-reference/chat/create#chat/create-functions)). But if we are going to talk about vulnerabilities, com.squareup.retrofit2:converter-jackson:2.9.0 already has some.

Feel free to update anything you need when doing the pull. I'm not a GitHub expert, but from what I know I think you can decide what to include and what not. I changed the .md and added the sections you commented with '?', so that people looking at my fork could install it on their own computer and use it instantly. You can leave only the necessary changes in the code.

If that's not possible, let me know, and I'll help, once again, with the necessary changes.

Eduardo-T commented:
Hello again @AndrewCPU,

After doing some research, I found that you could indeed make some changes, but it was tedious. So I already made them myself.

I've also added information on how to use the new feature in the README. Tell me what you think.

DanyBv commented:
Hi everyone, just to mention that this implementation does not seem to support streaming. (Choices should have a delta, etc.)

Grogdunn commented:
@Eduardo-T check this out https://github.com/optionfactory/openai-java/commit/024ecb9d5221a28d393ffbe5b8be09ea8a8a114b

on the documentation function_call (on top level) should be a string ["auto", "none"] ]or an object (containing a string :face_exhaling: )
```
Controls how the model responds to function calls. "none" means the model does not call a function, and responds to the end-user. "auto" means the model can pick between an end-user or calling a function. Specifying a particular function via {"name":\ "my_function"} forces the model to call that function. "none" is the default when no functions are present. "auto" is the default if functions are present.
```

Grogdunn commented:
Thx @Eduardo-T ported your changes to my fork :+1:

lumiseven commented:
@Eduardo-T 
Thank you for your pr, I tried it and it works very well. 
But if I use Stream mode(Stream=True), the type of `choices.0.delta.function_call.arguments` is in plain text format, so it would be better to use `StringSerializer` and `StringDeserializer` for the `getArguments` method in [ChatFunctionCallMixIn](https://github.com/Eduardo-T/openai-java/blob/fd91f5e670bfcb7668179295c1c70757f23a7722/service/src/main/java/com/theokanning/openai/service/ChatFunctionCallMixIn.java)?

davidb-e4s commented:
Just opened a PR to allow your changes to work with Kotlin @Eduardo-T

Eduardo-T commented:
> @Eduardo-T
> Thank you for your pr, I tried it and it works very well.
> But if I use Stream mode(Stream=True), the type of `choices.0.delta.function_call.arguments` is in plain text format, so it would be better to use `StringSerializer` and `StringDeserializer` for the `getArguments` method in [ChatFunctionCallMixIn](https://github.com/Eduardo-T/openai-java/blob/fd91f5e670bfcb7668179295c1c70757f23a7722/service/src/main/java/com/theokanning/openai/service/ChatFunctionCallMixIn.java)?

Thanks! Yes, I am currently working on an update that allows stream completions. I will be sharing it soon.

Eduardo-T commented:
> @Eduardo-T
> Thank you for your pr, I tried it and it works very well.
> But if I use Stream mode(Stream=True), the type of `choices.0.delta.function_call.arguments` is in plain text format, so it would be better to use `StringSerializer` and `StringDeserializer` for the `getArguments` method in [ChatFunctionCallMixIn](https://github.com/Eduardo-T/openai-java/blob/fd91f5e670bfcb7668179295c1c70757f23a7722/service/src/main/java/com/theokanning/openai/service/ChatFunctionCallMixIn.java)?

I implemented the solution. You can refer to "OpenAiApiFunctionsWIthStreamExample.java" to check an example.

Basically, for practical purposes, the stream can be mapped to a "ChatMessageAccumulator" which will append the function's and message's chunks, bringing everything together without you having to deal with it manually.

The changes now also take care of the "plain text format" issue you mentioned, as each chunk is converted to a TextNode, and when finished, the accumulated version is converted to a JsonNode which is most likely an ObjectNode. In general, you shouldn't even care about this, unless you're creating your own executor or something.

So my general recommendation would be to have a FunctionExecutor, and let it do the dirty work for you. For example:

Using FunctionExecutor, configure the executor of your function:
"functionBuilder.executor(Weather.class, w -> new WeatherResponse(w.location, w.unit, 25, "sunny"))"

And then just use:
"WeatherResponse functionExecutionResponse = functionExecutor.execute(chatFunctionCall);"

Or even better:
"ChatMessage callResponse = functionExecutor.executeAndConvertToMessageHandlingExceptions(chatFunctionCall);"

That way you can just add it to the conversation right away without even worrying about exceptions.

Eduardo-T commented:
> @Eduardo-T Thanks for doing this! I merged another PR that created some small conflicts, and can you add a basic test for this? Thanks again

No problem! It is done.

It now supports functions, functions with streaming, has tests and some documentation on some classes and also information in the README.md.

shaun-wild commented:
I'm a fan of the function executing solution you've come up with here, but personally I don't feel like it fits within the premise of the openai-java wrapper, it feels to me like its doing too much and would be better suited as a separate library.

Eduardo-T commented:
> I'm a fan of the function executing solution you've come up with here, but personally I don't feel like it fits within the premise of the openai-java wrapper, it feels to me like its doing too much and would be better suited as a separate library.

Thank you!

I don't think it's much, plus it's inside the 'service' component, and it's optional to use some components like FunctionExecutor. From then on, generally everything is necessary to have an optimal and easy operation related to the new feature of functions. And I repeat, it is in the 'service' module.

Adding one more module to the project I feel that it would add complexity, although I am interested in your proposal, could you elaborate?

aaronuu commented:
I would like to know when it can be merged into the main line
